lib/guard/api.rb

```diff
@@ -0,0 +1,279 @@
+require "guard"
+require "guard/internals/groups"
+
+module Guard
+  module API
+    TEMPLATE_FORMAT = "%s/lib/guard/%s/templates/Guardfile"
+
+    require "guard/ui"
+
+    def self.included(base)
+      base.extend(ClassMethods)
+      base.class_eval do
+        attr_accessor :group, :watchers, :callbacks, :options
+      end
+    end
+
+    module ClassMethods
+      # Get all callbacks registered for all Guard plugins present in the
+      # Guardfile.
+      #
+      def callbacks
+        @callbacks ||= Hash.new { |hash, key| hash[key] = [] }
+      end
+
+      # Add a callback.
+      #
+      # @param [Block] listener the listener to notify
+      # @param [Guard::Plugin] guard_plugin the Guard plugin to add the callback
+      # @param [Array<Symbol>] events the events to register
+      #
+      def add_callback(listener, guard_plugin, events)
+        Array(events).each do |event|
+          callbacks[[guard_plugin, event]] << listener
+        end
+      end
+
+      # Notify a callback.
+      #
+      # @param [Guard::Plugin] guard_plugin the Guard plugin to add the callback
+      # @param [Symbol] event the event to trigger
+      # @param [Array] args the arguments for the listener
+      #
+      def notify(guard_plugin, event, *args)
+        callbacks[[guard_plugin, event]].each do |listener|
+          listener.call(guard_plugin, event, *args)
+        end
+      end
+
+      # Reset all callbacks.
+      #
+      # TODO: remove (not used anywhere)
+      def reset_callbacks!
+        @callbacks = nil
+      end
+
+      # Returns the non-namespaced class name of the plugin
+      #
+      #
+      # @example Non-namespaced class name for Guard::RSpec
+      #   Guard::RSpec.non_namespaced_classname
+      #   #=> "RSpec"
+      #
+      # @return [String]
+      #
+      def non_namespaced_classname
+        to_s.sub("Guard::", "")
+      end
+
+      # Returns the non-namespaced name of the plugin
+      #
+      #
+      # @example Non-namespaced name for Guard::RSpec
+      #   Guard::RSpec.non_namespaced_name
+      #   #=> "rspec"
+      #
+      # @return [String]
+      #
+      def non_namespaced_name
+        non_namespaced_classname.downcase
+      end
+
+      # Specify the source for the Guardfile template.
+      # Each Guard plugin can redefine this method to add its own logic.
+      #
+      # @param [String] plugin_location the plugin location
+      #
+      def template(plugin_location)
+        File.read(format(TEMPLATE_FORMAT, plugin_location, non_namespaced_name))
```

Comment:
Can we do something better than this? Rather than putting templates in `lib/...`, which is a tiny bit odd since it's not code - why not put it in `templates/` of root of gem?